### PR TITLE
Hive: test dumper with an in-memory Hive metastore

### PR DESCRIPTION
**Pull request by paolomorandini**
Command:
```
./gradlew :dumper:app:{cleanTest,test} --tests HiveMetadataConnectorTest -Dtest.verbose=true -Dorg.gradle.java.home=/usr/lib/jvm/java-1.8.0-openjdk-amd64
```

### CODE REVIEW

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hive/AbstractHiveConnector.java [hunk 1]
```diff
@@ -147,7 +148,8 @@ public HiveMetastoreThriftClient newClient(@Nonnull String name) throws TTranspo
         @Nonnull
         public ThriftClientPool newMultiThreadedThriftClientPool(@Nonnull String name) {
             LOG.debug("Creating a new multi-threaded pooled Thrift client named '{}' backed by a thread pool of size {}.", name, threadPoolSize);
-            return new ThriftClientPool(name, thriftClientBuilder, ExecutorManager.newExecutorServiceWithBackpressure(name, threadPoolSize));
+            ExecutorService executorService = ExecutorManager.newExecutorServiceWithBackpressure(name, threadPoolSize);
+            return new ThriftClientPool(name, thriftClientBuilder, new ExecutorManager(executorService));
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hive/AbstractHiveConnector.java:152]: I think just pass in the ExecutorManager here.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hive/AbstractHiveConnector.java [hunk 2]
```diff
@@ -64,21 +61,25 @@ public abstract class AbstractHiveConnector extends AbstractConnector {
...
         private final Object lock = new Object();
         @GuardedBy("lock")
         @Nonnull
         private final List<@NonNull HiveMetastoreThriftClient> builtClients = new ArrayList<>();
 
-        public ThriftClientPool(@Nonnull String name, @Nonnull HiveMetastoreThriftClient.Builder thriftClientBuilder, @Nonnull ExecutorService executorService) {
+        public ThriftClientPool(@Nonnull String name, @Nonnull HiveMetastoreThriftClient.Builder thriftClientBuilder, @Nonnull ExecutorManager executorManager) {
             this.name = Preconditions.checkNotNull(name, "name was null.");
-            this.executorService = Preconditions.checkNotNull(executorService, "executorService was null.");
+            this.executorManager = Preconditions.checkNotNull(executorManager, "executorManager was null.");
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hive/AbstractHiveConnector.java:82]: This should create an ExecutorManager and keep it private. The caller doesn't need to know about it.

File: dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/hive/support/HiveServerSupport.java [hunk 3]
```diff
@@ -241,37 +208,56 @@ private void waitForState(HiveServer2 hiveServer, STATE targetState) throws Inte
...
 
   @Override
   public void close() throws Exception {
     hiveServer.stop();
     waitForState(hiveServer, STATE.STOPPED);
     metastoreExecutor.shutdown();
-    metastoreExecutor.awaitTermination(1, TimeUnit.MINUTES);
+    if (!metastoreExecutor.awaitTermination(1, TimeUnit.MINUTES)) {
+      LOG.info("Executor timed out.");
+    }
```
Comment (shevek-google) [inline dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/hive/support/HiveServerSupport.java:261]: Should this use MoreExecutors.shutdownAndAwaitTermination()?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hive/AbstractHiveConnector.java [hunk 4]
```diff
@@ -94,22 +95,22 @@ public ThriftClientPool(@Nonnull String name, @Nonnull HiveMetastoreThriftClient
...
 
         @Override
-        public void close() throws IOException {
-            LOG.debug("Closing pooled Thrift client '{}'.", name);
-            final int TIMEOUT = 30;
-            LOG.debug("Shutting down thread pool backing pooled Thrift client '{}'; will wait up to {} seconds", name, TIMEOUT);
-            MoreExecutors.shutdownAndAwaitTermination(executorService, TIMEOUT, TimeUnit.SECONDS);
+        public void close() throws Exception {
+            LOG.debug("Shutting down thread pool backing pooled Thrift client '{}'", name);
+            executorManager.close();
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hive/AbstractHiveConnector.java:113]: This also needs to shut down the executor, still. ExecutorManager doesn't do that. It only guarantees that any tasks submitted to it are complete.

**Review by shevek-google — DISMISSED** [review]
Very close, thank you.

File: dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/hive/support/HiveServerSupport.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright 2022 Google LLC
+ * Copyright 2013-2021 CompilerWorks
+ * Copyright (C) 2015-2021 Expedia, Inc. and Klarna AB.
```
Comment (paolomorandini) [inline dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/hive/support/HiveServerSupport.java:4]: This needed an exclusion from spotless.

**Review by bdw-g — APPROVED** [review]
Cool... am I understanding correctly that you believe this also fixes the immediate bug the customer was facing?
Approving, but subject to clarification about the license attribution, since I don't know htat.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hive/AbstractHiveConnector.java [hunk 6]
```diff
@@ -64,21 +63,28 @@ public abstract class AbstractHiveConnector extends AbstractConnector {
...
         private final Object lock = new Object();
         @GuardedBy("lock")
         @Nonnull
         private final List<@NonNull HiveMetastoreThriftClient> builtClients = new ArrayList<>();
 
-        public ThriftClientPool(@Nonnull String name, @Nonnull HiveMetastoreThriftClient.Builder thriftClientBuilder, @Nonnull ExecutorService executorService) {
+        public ThriftClientPool(@Nonnull String name, @Nonnull HiveMetastoreThriftClient.Builder thriftClientBuilder, int threadPoolSize) {
             this.name = Preconditions.checkNotNull(name, "name was null.");
-            this.executorService = Preconditions.checkNotNull(executorService, "executorService was null.");
+            this.executorService = ExecutorManager.newExecutorServiceWithBackpressure(name, threadPoolSize);
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hive/AbstractHiveConnector.java:86]: Interesting. I might have still passed in the ExecutorService, as things like caller-thread-only/blocking ExecutorService(s) exist, and are useful for testing. Actually, this isn't critical, as newESWB returns a DirectExecutor if threadPoolSize is 1.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java [hunk 7]
```diff
@@ -77,7 +75,11 @@ public JdbcPropBuilder prop(@Nonnull String prop, @Nonnull String val) {
     }
 
     private void addProp(String prop, String val) {
-        props.add(prop + punctuations.charAt(1) + URLEncoder.encode(val, UTF_8));
+        try {
+            props.add(prop + punctuations.charAt(1) + URLEncoder.encode(val, "UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java:81]: This should go to a logger at the very least. I wonder how the InternalError above sneaked past review!

**Review by shevek-google — DISMISSED** [review]
Sorry, changes. Mostly for the printStackTrace() instead of a slf4j logger.

File: dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/hive/support/HiveServerSupport.java [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+  @Positive
+  public int getMetastoreThriftPort() {
+    return thriftPortMetastore;
+  }
+
+  @SuppressWarnings("deprecation")
+  public void configure() {
+    try {
+      hiveTestDir = Files.createTempDirectory("dumper-hive-test-");
```
Comment (paolomorandini) [inline dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/hive/support/HiveServerSupport.java:92]: Needs to be deleted on `close()`

File: dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/hive/HiveMetadataConnectorTest.java [hunk 9]
```diff
@@ -16,23 +16,149 @@
...
- * @author swapnil
+ * Run this with: > ./gradlew :dumper:app:{cleanTest,test} --tests HiveMetadataConnectorTest
+ * -Dtest.verbose=true -Dorg.gradle.java.home=/usr/lib/jvm/java-1.8.0-openjdk-amd64
  */
 @RunWith(JUnit4.class)
 public class HiveMetadataConnectorTest extends AbstractConnectorTest {
 
-    private final HiveMetadataConnector connector = new HiveMetadataConnector();
+  private static final Logger LOG = LoggerFactory.getLogger(HiveMetadataConnectorTest.class);
+  private final boolean debug = true;
```
Comment (paolomorandini) [inline dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/hive/HiveMetadataConnectorTest.java:59]: Set to `false`

File: gradle.properties [hunk 10]
```diff
@@ -0,0 +1,3 @@
+org.gradle.jvmargs=-Xmx2g -Dfile.encoding=UTF-8 -XX:+HeapDumpOnOutOfMemoryError
```
Comment (paolomorandini) [inline gradle.properties:1]: Default `Xmx` fails consistently on JDK 8.

**Review by shevek-google — APPROVED** [review]

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java [hunk 11]
```diff
@@ -77,7 +75,11 @@ public JdbcPropBuilder prop(@Nonnull String prop, @Nonnull String val) {
     }
 
     private void addProp(String prop, String val) {
-        props.add(prop + punctuations.charAt(1) + URLEncoder.encode(val, UTF_8));
+        try {
+            props.add(prop + punctuations.charAt(1) + URLEncoder.encode(val, "UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            LOG.warn("Could not encode {}={} into JDBC URL, the property will be dropped.", prop, val);
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java:81]: This looks like a semantic change. We called a method, and it did nothing but log the fact that it didn't do what was asked. Why do we need this? What invalid parameter are we trying to pass and ignore?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java [hunk 12]
```diff
@@ -77,7 +75,11 @@ public JdbcPropBuilder prop(@Nonnull String prop, @Nonnull String val) {
     }
 
     private void addProp(String prop, String val) {
-        props.add(prop + punctuations.charAt(1) + URLEncoder.encode(val, UTF_8));
+        try {
+            props.add(prop + punctuations.charAt(1) + URLEncoder.encode(val, "UTF-8"));
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java:79]: Do we prefer the string literal to the symbolic constant? Is this a JDK support/availability question? StandardCharsets is available on JDK7.

**Review by shevek-google — CHANGES_REQUESTED** [review]

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java (hunk 12 above)
Comment (paolomorandini) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java:79]: It's the `URLEncoder.encode` overload which is missing in jdk 8.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java (hunk 11 above)
Comment (paolomorandini) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/redshift/JdbcPropBuilder.java:81]: Right, probably just throwing and make the dumper fail would make more sense.

**Pull request merged**